### PR TITLE
Wc upwind fix

### PR DESCRIPTION
**Pull request by chenwilliam77**
Implement correct upwinding (on the interior nodes but not boundary conditions yet). Add examples from SimpleDifferentialOperators. Add additional tests for matrix_free_operators.jl. Add additional tests for both uniform and non-uniform stencils for the CentralDifference and UpwindDifference operators.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Is this WIP?

**Comment by chenwilliam77** [comment]
> Is this WIP?

Yes, Adam and I are still coordinating on fixing upwinding. Just to check that I had created this pull request correctly, I was intending to pull request Adam's branch upwind_fix. I did do this correctly, right?

**Comment by ajozefiak** [comment]
> I was intending to pull request Adam's branch upwind_fix. I did do this correctly, right?

Not quite, you're making a pull request into master, but it's fine for now since we're still finishing everything up.

**Comment by chenwilliam77** [comment]
Hi @ChrisRackauckas, just letting you know that we are ready to merge the PR if the changes we have made seem fine with you. Let @ajozefiak or me know if you have questions or additional requests for changes before merging the PR.

**Comment by ajozefiak** [comment]
> Hi @ChrisRackauckas, just letting you know that we are ready to merge the PR if the changes we have made seem fine with you. Let @ajozefiak or me know if you have questions or additional requests for changes before merging the PR.

Hold on a moment, this branch is still missing the last 4 commits of my branch, which are necessary for convolutions against a boundary padded vector.

**Comment by ajozefiak** [comment]
Actually, I'm okay merging this. We can make a followup PR for boundary padded vectors, otherwise this massive change will be drawn out.

File: src/derivative_operators/abstract_operator_functions.jl [hunk 1]
```diff
@@ -164,7 +166,12 @@ end
 ################################################################################
 
 function *(coeff_func::Function, A::DerivativeOperator{T,N,Wind}) where {T,N,Wind}
-    coefficients = A.coefficients === nothing ? Vector{T}(undef,A.len) : A.coefficients
+    if hasmethod(coeff_func, (Float64,)) || hasmethod(coeff_func, (Vector{Float64},))
```
Comment (ChrisRackauckas) [inline src/derivative_operators/abstract_operator_functions.jl]: Why Float64? That's not gonna fly. Also, `hasmethod` will be really fragile, do not rely on it.

File: src/derivative_operators/coefficient_functions.jl [hunk 2]
```diff
@@ -0,0 +1,67 @@
+"""
+```
+compute_coeffs(coeff_func, current_coeffs)
+```
+calculates the coefficients for the stencil of UpwindDifference operators.
+"""
+function compute_coeffs!(coeff_func::Number, current_coeffs::AbstractVector{T}) where {T<:Number}
+    return current_coeffs[:] .+= coeff_func
```
Comment (ChrisRackauckas) [inline src/derivative_operators/coefficient_functions.jl]:
```suggestion
    return current_coeffs .+= coeff_func
```

File: src/matrixfree_operators.jl [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import LinearAlgebra: mul!
+import Base.==
```
Comment (ChrisRackauckas) [inline src/matrixfree_operators.jl]: should not do imports like this. In fact, that mul! one should be deleted too

**Comment by ChrisRackauckas** [comment]
let's merge because it's better, and fix up that coefficient stuff first.
